python/ray/data/extensions/tensor_extension.py

```diff
@@ -1282,14 +1282,21 @@ def _to_numpy(self,
         storage_list_type = self.storage.type
         ext_dtype = storage_list_type.value_type.to_pandas_dtype()
         shape = self.type.shape
-        # Size in bytes of the underlying ndarray item.
-        item_byte_width = storage_list_type.value_type.bit_width // 8
+        value_type = storage_list_type.value_type
+        if pa.types.is_boolean(value_type):
+            # Boolean array buffers are byte-packed, with 8 entries per byte,
+            # and are accessed via bit offsets.
+            buffer_item_width = value_type.bit_width
+        else:
+            # We assume all other array types are accessed via byte array
+            # offsets.
+            buffer_item_width = value_type.bit_width // 8
```

Comment:
Could we get this magic number "8" programmatically from Arrow instead of special casing?

Reply:
I tried to find a way to do that, but this is fundamentally a value type question: boolean arrays are byte-packed and use bit offsets (see [Pandas writer](https://github.com/apache/arrow/blob/415439cd0a30732830249c6924dff945b88630b4/cpp/src/arrow/python/arrow_to_pandas.cc#L1317-L1327) and [indexing logic](https://github.com/apache/arrow/blob/e903a214525dd6dcd8e57b20958a65dc678be47d/cpp/src/arrow/array/array_primitive.h#L50-L53)), all other primitive type arrays aren't byte-packed and [use byte offsets](https://github.com/apache/arrow/blob/415439cd0a30732830249c6924dff945b88630b4/cpp/src/arrow/python/arrow_to_pandas.cc#L282-L284). I haven't found a way to reliably determine whether a type's buffer is byte-addressable or bit-addressable.

Reply:
The magic number "8" is just `CHAR_BIT`, the number of bits in a `char` byte, which is 8 on all platforms that we wish to support (`CHAR_BIT == 8` on all POSIX systems and Windows). I couldn't find a way to programmatically retrieve `CHAR_BIT` from Python, the closest I could find was `os.sysconf("SC_CHAR_BIT")`, but that's a GNU extension, not available on all non-Unix platforms nor even all Unix platforms.

I can also define a module-level `NUM_BITS_IN_BYTE = 8` constant if you feel that this magic number is too opaque?

Reply:
> all other primitive type arrays aren't byte-packed and use byte offsets.

Thanks, this was what I wanted to know.